doc/developer/command_and_response_binary_encoding.md

```diff
@@ -0,0 +1,574 @@
+2. Implement a pair mappings that convert between `$T` and `Proto$T`.
+
+## Overview
+
+As a running example, consider a new type located in `mz_expr::foo::bar::$T`.
+In order to add Protobuf support for `$T`, you have to change the the source code of the `mz_expr` crate as follows.
+
+- `expr` - crate root folder.
+  - `build.rs` - contains `prost_build` instructions for compiling all `*.proto` files in the crate into `*.rs` source code.
+    - [Add instructions for compiling `foo/bar.proto`.](#extending-buildrs)
+  - `src` - crate sources folder.
+    - `foo/bar.proto` - contains Protobuf definitions `Proto$T` for types `$T` located in `foo/bar/mod.rs`.
+      - [Add a Protobuf message definition for `Proto$T`.](#defining-a-protobuf-message-for-protot)
+    - `foo/bar/mod.rs` - contains Rust definitions `$T` and `Proto$T` and associated traits.
+      - [Add the Rust definition of `$T`.](#defining-a-rust-typet)
+      - [Include the Rust code of the generated `Proto$T` type.](#including-rust-sources-generated-by-prost)
+      - [Implement `$T ⇔ Proto$T` mappings.](#implementing-t--protot-mappings)
+      - [Add unit tests for `$T`.](#adding-unit-tests-for-t)
+
+The following sections contain details for each action item listed above.
+
+## Defining a Rust type`$T`
+
+We consider two main cases for `$T` - structs and enums.
+Here are the two definitions from `expr/src/foo/bar/mod.rs` to be used as a running example.
+
+```rust
+use chrono::NaiveDate;
+use mz_repr::adt::char::CharLength;
+
+// `$T` is a struct
+#[derive(Debug)]
+pub struct MyStruct {
+    pub field_1: u64,
+    pub field_2: usize,
+    pub field_3: CharLength,
+    pub field_4: NaiveDate,
+}
+
+// `$T` is an enum
+#[derive(Debug)]
+pub enum MyEnum {
+    Var1(u64),
+    Var2(usize),
+    Var3(CharLength),
+    Var4(NaiveDate),
+}
+```
+
+The above examples also illustrate of the classes of nested Rust types that one may encounter<a name="type-classes"></a>:
+
+1. Primitive types that have a Protobuf counterpart (such as `u64`).
+2. Primitive types that don't have a Protobuf counterpart (such as `usize`).
+3. Complex types that are defined by us (such as `MyLibType`).
+4. Complex types that are not defined by us (such as `DateTime`).
+
+The problem of encoding `$T` in a Protobuf-based binary format thereby decomposes into the problem of encoding instance of each of the above four classes.
+The following rules apply in general:
+
+- Types from (1) are trivially represented by their existing Protobuf counterpart.
+- Types from (3) are the primary focus of this guide.
+- Types from (2) only care about implementation steps (3) and (4), and types that belong to (4) don't care about step (1).
+- Types from (2) and (4) require a slightly different implementation mechanism for steps (3) and (4) compared to types from (3).
+  The [`chrono` types PR](https://github.com/MaterializeInc/materialize/pull/11801) is a good example of the latter.
```

Comment:
I find the mention of "steps" to be confusing because 1) the steps themselves are not numbered. 2) there are only 4 steps mentioned, but there are 5 headers from this point on. I would guess that  "Including Rust sources generated by `prost`" does not count as a "step".

Also, I prefer combining the two lists. For example, the first two items on the list can read, 
"1. Primitive types that have a Protobuf counterpart (such as `u64`). They are trivially represented by their protobuf counterpart." 
"2. Primitive types that don't have a Protobuf counterpart (such as `usize`). We do not have to define a new Protobuf message for the primitive type, but we do need to define a mapping from the primitive type to some Protobuf scalar type."

Reply:
> I find the mention of "steps" to be confusing because 1) the steps themselves are not numbered.
They are, they are referring to the three steps at the top of the document (they were originally four, but I merged the last two). I added hyperlinks in the text to make the relationship more explicit.

Markdown does not allow for alphanumeric lists, but I think I can use inline HTML to make the type classes list by letters (so the doc will be talking about class A", B, etc.